README.md
# bilateral-filter
Bilateral filter on single channel floating point image.


history.txt
a
a.shape
b = a.view((np.uint16, 2))
b = a.view((np.uint16, 1))
b.shape
import cv2
cv2.imwrite("test.jp2", b)
set OPENCV_IO_ENABLE_JASPER=ON
set OPENCV_IO_ENABLE_JASPER:ON

## ---(Wed Oct 16 13:59:23 2019)---
import glymur, numpy as np
pwd
cd Development/
cd bilateral-filter/
pwd
jp2 = glymur.Jp2k('zeros.jp2', data=np.zeros((640, 480), dtype=np.uint8))
jp2 = glymur.Jp2k('zeros.jp2', data=np.zeros((640, 480), dtype=np.float32))
jp2 = glymur.Jp2k('zeros.jp2', data=np.zeros((640, 480), dtype=np.uint16))
a = run_bilateral_filter('Z_l3.png', 3, 0.01, 0.5)
from bilateral_filter import run_bilateral_filter
a = run_bilateral_filter('Z_l3.png', 3, 0.01, 0.5)
b = a.view((np.uint16, 2))
b = a.view((np.uint16, 1))
b
jp2 = glymur.Jp2k('zeros.jp2', b)
b.shape
jp2 = glymur.Jp2k('zeros.jp2', b[:,:,0])
jp2 = glymur.Jp2k('zeros2.jp2', b[:,:,1])
import cv2
im = cv2.imread('Z_l0.png', cv2.IMREAD_UNCHANGED)
im
im.shape
jp2 = glymur.Jp2k('Z_l0_converted.jp2', im)
r = im[:,:,0]
im[:,:,0] = im[:,:,1]
im[:,:,1] = r
jp2 = glymur.Jp2k('Z_l0_converted.jp2', im)
from glymur.core import COLOR, OPACITY
ctype = [COLOR, COLOR, COLOR, OPACITY]
from glymur.core import RED, GREEN, BLUE, WHOLE_IMAGE
asoc = [RED, GREEN, BLUE, WHOLE_IMAGE]
cdef = glymur.jp2box.ChannelDefinitionBox(ctype, asoc)
print(cdef)
boxes = jp2.box
boxes[2].box.append(cdef)
jp2_rgba = jp2.wrap("goodstuff_rgba.jp2", boxes=boxes)
inim = cv2.imread('Z_l0.png', cv2.IMREAD_UNCHANGED)
inim = inim.view((np.uint8, 4))
inim.shape
jp2 = glymur.Jp2k('Z_l0_converted.jp2', inim)
boxes = jp2.box
boxes[2].box.append(cdef)
jp2_rgba = jp2.wrap("goodstuff_rgba.jp2", boxes=boxes)
cv2.imwrite('test.png', inim)
jp2 = glymur.Jp2k('Z_l0_converted_0.jp2', inim[:,:,0])
jp2 = glymur.Jp2k('Z_l0_converted_0.jp2', inim[:,:,1])
jp2 = glymur.Jp2k('Z_l0_converted_0.jp2', inim[:,:,0])
jp2 = glymur.Jp2k('Z_l0_converted_1.jp2', inim[:,:,1])
jp2 = glymur.Jp2k('Z_l0_converted_2.jp2', inim[:,:,2])
jp2 = glymur.Jp2k('Z_l0_converted_3.jp2', inim[:,:,3])
a = run_bilateral_filter('Z_l3.png', 3, 0.01, 1)
a = run_bilateral_filter('Z_l3.png', 3, 0.1, 1)
a = run_bilateral_filter('Z_l3.png', 3, 0.01, 1.5)
a = run_bilateral_filter('Z_l3.png', 3, 0.001, 2)
a = run_bilateral_filter('Z_l3.png', 3, 2, 2)
a = run_bilateral_filter('Z_l0_crop.png', 3, 0.1, 2)
inim = cv2.imread('Z_l0_crop.png', cv2.IMREAD_UNCHANGED)
inim.shape
inim
f = inim.view((np.float32,1))
f
inim
a = cv2.imwread('Z_L0.png', cv2.IMREAD_UNCHANGED)
a = cv2.imread('Z_L0.png', cv2.IMREAD_UNCHANGED)
a
a.shape
b = a[1:200,1:200,:]
b
cv2.imwrite('Z_l0_crop.png', b)
b = a[1:500,1:500,:]
cv2.imwrite('Z_l0_crop.png', b)
a = run_bilateral_filter('Z_l0_crop.png', 3, 0.1, 2)
blur = cv2.blur(b,(5,5))
cv2.imwrite('Z_l0_crop_blur5.png', blur)
blur = cv2.blur(b,(3,3))
cv2.imwrite('Z_l0_crop_blur5.png', blur)
cv2.imwrite('Z_l0_crop_blur3.png', blur)
blur = cv2.blur(b,(5,5))
cv2.imwrite('Z_l0_crop_blur5.png', blur)
gblur = cv2.GaussianBlur(b,(5,5),0)
cv2.imwrite('Z_l0_crop_gblur5.png', gblur)
gblur = cv2.GaussianBlur(b,(3,3),0)
cv2.imwrite('Z_l0_crop_gblur3.png', gblur)
cv2.imwrite('Z_l0_crop_blur3.png', blur)cd ..
cd ..
cd EclConverter/
ConvertEclToImages.py
runfile('C:/Users/sicatrb/Development/EclConverter/ConvertEclToImages.py', wdir='C:/Users/sicatrb/Development/EclConverter')
pwd
runfile('C:/Users/sicatrb/Development/EclConverter/ConvertEclToImages.py', wdir='C:/Users/sicatrb/Development/EclConverter')

## ---(Wed Oct 16 15:33:28 2019)---
runfile('C:/Users/sicatrb/Development/EclConverter/ConvertEclToImages.py', wdir='C:/Users/sicatrb/Development/EclConverter')
runfile('C:/Users/sicatrb/Development/EclConverter/ConvertEclToImages.py', wdir='C:/Users/sicatrb/Development/EclConverter')
runfile('C:/Users/sicatrb/Development/EclConverter/ConvertEclToImages2.py', wdir='C:/Users/sicatrb/Development/EclConverter')
from ecl import EclGrid
from ecl.grid import EclGrid
pwd
cd ..
cd bilateral-filter/
import cv2
im = cv2.imread('Z_l0_crop.png', cv2.IMREAD_UNCHANGED)
im
imf = im.view((np.float32, 1))
import numpy as np
imf = im.view((np.float32, 1))
imf
imf = imf * 5000
imf
imf = imf + 5000
imf
imf = imf * 1000
imf
import math
imf = floor(imf)
imf = math.floor(imf)
imf = imf * 10
imf
cv2.imwrite('imf.png', imf.view((np.uint8, 4)))
imf.shape
imu = imf.view((np.uint8, 4))
imu
imu.shape
imu = imu[:,:,0,:]
cv2.imwrite('imf.png', imu)
imf
im = cv2.imread('Z_l0_crop.png', cv2.IMREAD_UNCHANGED)
imf = im.view((np.float32, 1))
imf
imf = imf + 1
imf
imf = imf * 1000000000
imf
imf = imf / 10
imf
imf = imf / 10
imf
imf = im.view((np.float32, 1))
imf
imf = imf + 1
imf = imf * 5000000
imf
imf = imf.view((np.int32, 1))
imf
imf = im.view((np.float32, 1))
imf
imf = imf * 500000
imf = imf + 500000
imf
imf.round(0)
imf
imf = imf.round(0)
imf
imf.dtype
imf = im.view((np.float32, 1))
imf.dtype
imf = imf * 50000
imf.dtype
imf = imf + 50000
imf.dtype
imf
imf = imf.round(0)
imf
cv2.imwrite('imf.png', imf)
imb = imf.view((np.uint8, 4))
imb
img.shape
imb.shape
imb = imb[:,:,0,:]
imb = imf.view((np.uint8, 4))
imb = imb[:,:,0,:]
cv2.imwrite('imf.png', imb)
imbf = imb.view((np.float32, 1))
imbf
imbf = imbf - 50000
imbf = imbf / 50000
imbf
imbfu = imbf.view((np.uint8, 4))
imbfu.shape
imbfu = imbfu[:,:,0,:]
cv2.imwrite('imbfu.png', imbfu)
imbf
imf = cv2.imread('Z_l0_crop.png', cv2.IMREAD_UNCHANGED)
imf
imf = imf.view((np.float32, 1))
imf
a = cv2.imread('Z_L3.png', cv2.IMREAD_UNCHANGED)
b = cv2.imread('filtered_Z_l3.png_d3_si2_ss2.png', cv2.IMREAD_UNCHANGED)
a
a = a.view((np.float32, 1))
b = b.view((np.float32, 1))
a
b
b = cv2.imread('filtered_Z_l3.png_d3_si0.01_ss1.png', cv2.IMREAD_UNCHANGED)
b = b.view((np.float32, 1))
b
a
a = cv2.imread('Z_L0.png', cv2.IMREAD_UNCHANGED)
b = cv2.imread('filtered_Z_l0.png_d3_si0.01_ss0.5.png', cv2.IMREAD_UNCHANGED)
b = b.view((np.float32, 1))
a = a.view((np.float32, 1))
a
b
ai = cv2.imread('Z_l0_crop.png', cv2.IMREAD_UNCHANGED)
ai = ai.view((np.float32, 1))
ai
ai = ai * 100000
ai = ai + 100000
ai
ai.round(0)
ai.dtype

bilateral-filter.py
# -*- coding: utf-8 -*-
"""
Created on Tue Oct 15 13:54:04 2019

@author: sicatrb
"""

import cv2
import numpy as np
import sys
import math

filename = "l0.png"

image = cv2.imread(filename, cv2.IMREAD_UNCHANGED)

imagefloat = image.view((np.float32,1))

def distance(x, y, i, j):
    return np.sqrt((x-i)**2 + (y-j)**2)


def gaussian(x, sigma):
    return (1.0 / (2 * math.pi * (sigma ** 2))) * math.exp(- (x ** 2) / (2 * sigma ** 2))


def apply_bilateral_filter(source, filtered_image, x, y, diameter, sigma_i, sigma_s):
    hl = math.floor(diameter/2)
    i_filtered = 0
    Wp = 0
    i = 0
    while i < diameter:
        j = 0
        while j < diameter:
            neighbour_x = x - (hl - i)
            neighbour_y = y - (hl - j)
            if neighbour_x >= len(source):
                neighbour_x -= len(source)
            if neighbour_y >= len(source[0]):
                neighbour_y -= len(source[0])
            gi = gaussian(source[neighbour_x][neighbour_y] - source[x][y], sigma_i)
            gs = gaussian(distance(neighbour_x, neighbour_y, x, y), sigma_s)
            w = gi * gs
            i_filtered += source[neighbour_x][neighbour_y] * w
            Wp += w
            j += 1
        i += 1
    i_filtered = i_filtered / Wp
    filtered_image[x][y] = i_filtered


def bilateral_filter_own(source, filter_diameter, sigma_i, sigma_s):
    filtered_image = np.zeros(source.shape)

    i = 0
    while i < len(source):
        j = 0
        while j < len(source[0]):
            apply_bilateral_filter(source, filtered_image, i, j, filter_diameter, sigma_i, sigma_s)
            j += 1
        i += 1
    return filtered_image

filtered_image = bilateral_filter_own(imagefloat, 5, 12.0, 16.0)
cv2.imwrite("filtered_image.png", filtered_image.view((np.uint8, 4)))

if __name__ == "__main__":
    filename = str(sys.argv[1])
    src = cv2.imread(filename, cv2.IMREAD_UNCHANGED).view((np.float32, 1))
    
    #filtered_image_OpenCV = cv2.bilateralFilter(src, 5, 12.0, 16.0)
    #cv2.imwrite("original_image_grayscale.png", src)
    #cv2.imwrite("filtered_image_OpenCV.png", filtered_image_OpenCV)
    diameter = sys.argv[2]
    sigma_i = sys.argv[3]
    sigma_space = sys.argv[4]
    filtered_image = bilateral_filter_own(src, diameter, sigma_i, sigma_space)
    cv2.imwrite("filtered_" + filename  + "_d" + diameter + "_si" + sigma_i +
                "_ss" + sigma_space + ".png", filtered_image)

bilateral_filter.py
# -*- coding: utf-8 -*-
"""
Created on Tue Oct 15 13:54:04 2019

@author: sicatrb
"""

import cv2
import numpy as np
import math

def gaussianF(x, sigma):
    return (1.0 / (2 * math.pi * (sigma ** 2))) * math.exp(- (x ** 2) / (2 * sigma ** 2))

def gaussianApprox(x, sigma):
    
    alpha = 0.5 / sigma
    
    x *= alpha
    if(x < -2):
	    return 0
    if(x < -1):
	    x += 2
	    return x * x * x
    if(x < 0):
	    return (4 - 3*x*x*(2 + x))
    if(x < 1):
	    return (4 - 3*x*x*(2 - x))
    if(x < 2):
	    x = 2-x
	    return x * x * x
    return 0

def filter_pixel(source, filtered_image, x, y, diameter, sigma_i, sigma_s):
    hl = math.floor(diameter/2)
    c = 0
    i_filtered = 0
    Wp = 0
    i = 0
    while i < diameter:
        j = 0
        neighbour_x = x - (hl - i)
        if neighbour_x >= len(source):
            neighbour_x -= len(source)
                
        gx = gaussianApprox(neighbour_x - x, sigma_s)
        
        while j < diameter:
            
            neighbour_y = y - (hl - j)   
            if neighbour_y >= len(source[0]):
                neighbour_y -= len(source[0])
            gy = gaussianApprox(neighbour_y - y, sigma_s)    
             
            gi = gaussianApprox(source[neighbour_x][neighbour_y] - source[x][y], sigma_i)
            gs = gx * gy
            if(c == 0 and x == 10 and y == 10):       
                
                print("x,y = " + str(neighbour_x) + ", " + str(neighbour_y) + ": gi(" + str(source[neighbour_x][neighbour_y] - source[x][y]) + ") = " + str(gi) + ", gs = " + str(gs) + "\n")
            
            w = gi * gs
            i_filtered += source[neighbour_x][neighbour_y] * w
            Wp += w
            j += 1
        i += 1
    c += 1
    i_filtered = i_filtered / Wp
    #print("in: ")
    #print(i_filtered)
    #print("out: ")
    filtered_image[x][y] = i_filtered
    #print(filtered_image[x][y])
    #if source[x][y] != i_filtered:
    #    print("not equal")


def bilateral_filter(source, filter_diameter, sigma_i, sigma_s):
    filtered_image = np.zeros(source.shape, dtype=np.float32)

    i = 0
    while i < len(source):
        j = 0
        while j < len(source[0]):
            filter_pixel(source, filtered_image, i, j, filter_diameter, sigma_i, sigma_s)
            j += 1
        i += 1
    return filtered_image

def run_bilateral_filter(filename, diameter, sigma_i, sigma_space):
   
    src = cv2.imread(filename, cv2.IMREAD_UNCHANGED)
    src = src.view((np.float32, 1))
    
    print(src)
    
#    i = 0
#    while i < len(src):
#        j = 0
#        while j < len(src[0]):
#            src[i][j] = src[i][j] + 1.0
#            j += 1
#        i += 1
#    
    
    filtered_image = bilateral_filter(src, diameter, sigma_i, sigma_space)

#    i = 0
#    while i < len(src):
#        j = 0
#        while j < len(src[0]):
#            filtered_image[i][j] = filtered_image[i][j] - 1.0
#            j += 1
#        i += 1

    
    print(filtered_image)
    
    filtered_image = filtered_image.view((np.uint8, 4))
    filtered_image = filtered_image[:,:,0,:]
    #print(filtered_image)
    
    cv2.imwrite("filtered_" + filename  + "_d" + str(diameter) + "_si" + str(sigma_i) +
                "_ss" + str(sigma_space) + ".png", filtered_image)
    
    return filtered_image